<a href="https://colab.research.google.com/github/ManJ-PC/Py/blob/master/CursoEmV%C3%ADdeoMundo2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# if carro.esquerda():
#   bloco 1
# elif carro.direita():
#   bloco 2
# elif carro.ré():
#   bloco 3
# else:
#   bloco 4 NO

In [ ]:
# nome = str(input('Qual é o seu nome? '))
# if nome == 'Gustavo':
#   print('Que nome bonito!')
# elif nome in 'Ana Claúdia Jéssica Juliana':
#   print('Belo nome feminino')
# else:
#   print('Seu nome é bem normal.')
# print('Tenha um bom dia, {}!'.format(nome))

**Desafio  036**

Escreva um programa para aprovar o empréstimo bancário para a compra de uma casa. O Programa vai perguntar o valor da casa, o salário do comprador e em quantos anos ele vai pagar.

Calcule o valor da prestação mensal, sabendo que ela não pode exceder 30% do salário ou então o empréstimo será negado.

In [ ]:
# patrimonyL = float(input('Good Morning to our House Loan\'s Bank!\n Which the value of property that wants to adquire: '))
# wage = float(input('What is your wage: '))
# ageY = float(input('How much time lasts bank loan, in years: '))

# monthPay = patrimonyL/(ageY * 12)

# print('The mensal value to pay the house is: ', monthPay)

# if monthPay > 0.30 * wage:
#   print("Empréstimo negado")
# else:
#   print("Aceite!!")


In [ ]:
# #SOLUÇÂO

# casa = float(input('Valor da casa: R$  '))
# salario = float(input('Salário do comprador: R$  '))
# anos = int(input('Quantos anos de financiamento?  '))
# prestação = casa / (anos * 12)
# minimo = salario * 30 / 100 # min = .3*wage
# print('Para pagar uma casa de R${:.2f} em {} anos'.format(casa, anos), end='')
# """, end='')"""
# print(' a prestação será de R${:.2f}'.format(prestação))
# #print('Para pagar uma casa de R${:.2f}'.format(prestação))

# #print('COMPARANDO tem que pagar {} e o mínimo é de {}'.format(prestação,minimo))

# if prestação <= minimo: 
#    print('Empréstimo pode ser CONCEDIDO')
# else:
#   print('Empréstimo NEGADO')

**Desafio 37**

Escreva um programa que leia um número inteiro qualquer e peça para o usuário escolher qual será a **base de conversão**


*   1 para binário
*   2 para octal
*   3 para hexadecimal



In [ ]:
# num = int(input('Insira o valor decimal: '))
# b2 = bin(num)
# o8 = oct(num)
# Ox = hex(num)

# base = str(input('Qual será a base de conversão?\n O <- octal\n H <- hexadecimal\n B or other thing <- binary '))
# if base.upper() == 'O':
#   print(o8)
# elif base.upper() == 'H':
#   print(Ox)
# else:
#   print(b2)

In [ ]:
# SOLUÇÃO


# num = int(input('Digite um numero inteiro: '))
# print('''Escolha uma das bases para conversão:
# [ 1 ] converter para BINÁRIO
# [ 2 ] converter para OCTAL
# [ 3 ] converter para HEXADECIMAL''')
# opção = int(input('Sua opção: '))
# if opção == 1:
#   print('{} convertido para BINÁRIO é igual a {}'.format(num, bin(num)))
# elif opção == 2:
#   print('{} convertido para OCTAL é igual a {}'.format(num, oct(num)))
# elif opção == 3:
#   print('{} convertido para HEXADECIMAL é igual a {}'.format(num, hex(num)))

In [ ]:
# def converte2(a,b): #a = bq + r
#   novabase = "" #resto ou r_string
#   r = 0
#   while a>0: #r<a: #r > 0: ...n//b!
#     q = a // b
#     r = a - (q*b)
#     #r = a % b
#     a = q 

#     print(r,'\n')
#     #novabase=str(r)
#     novabase += str(r)
#   novabase = novabase[::-1]
#   #novabase[::-1]
#   #print('\n',novabase[::-1])
#   return novabase

In [ ]:
# def converte(n,b): #n = bq + r
#   # 10b10 = 1010b2  10/2 r1 5/2 r1 2/2 r0 0/2 r0
#   r_string = ""
#   while n//b > 0:
#     r = n%b
#     a = n//b
#     print(r,'\n')
#     r_string += str(r)
#   r_string[::-1]
#   print('')  NO
    

    


In [ ]:
# converte2(12,2)

In [ ]:
# converte2(12, 8)

In [ ]:
# Solution
def dec_to_base(num,base): # Maximum base - 36
  base_num = ""
  while num>0:
    dig = int(num%base) # porque este int? 
    if dig<10:
      base_num += str(dig)
    else:
      base_num += chr(ord('A')+dig-10) # Using uppercase Letters
    num //= base

  base_num = base_num[::-1] #To reverse the string
  return base_num

In [ ]:
dec_to_base(77,2)

In [ ]:
int("12",8)

In [ ]:
ord('a')

In [ ]:
ord('A')

In [ ]:
chr(94)+chr(93)+chr(92)+chr(91)+chr(90)

In [ ]:
# Any Base to Decimal

print(int("12",3))
print(int("12",8))
print(int("12",10))
print(int("12",16))

#chr(16)

In [ ]:
username = input("Choose a username: ")  # Canita Broa
final_username = username.replace("r", "_")

print("Your username is: " + final_username) 

In [ ]:
for element in final_username: #string_name:
  print(element, end=' ')
print("\n")

G a l o 



In [ ]:
def base_to_dec(num_str, base):
 base_num = 0
 power = 0           # declaration and initialization
 for column in num_str[::-1]: # d
   print(column)
   base_num += int(column) * pow(base, power)  # c = a + b 
   power += 1 # e
 print(base_num)
 #falta a potência!!! Incorporar, dessa forma em vez 12*8 = 96
# We have:

In [ ]:
num_str = "147"
num_str[::-1]

'741'

In [ ]:
base_to_dec("147",8)

7
4
1
103


In [ ]:
base_to_dec("147",8)

7
4
1
103


In [ ]:
print(pow(8,3))
7*64 # valores muito próximos, logicamente... pois base = 8 , 7 imediamente antes!

512


448

In [ ]:
base_to_dec("147",8)

7
4
1
103


In [ ]:
pow(4,4)

256

In [ ]:
pow(4,4,8)

0

In [ ]:
pow(4,4,12)

4

In [ ]:
# a,b,c = 100
# print(c)  .> cannot unpack non-iterable int object

a,b,c = 1,2,3
print(a,b,c)

1 2 3


In [ ]:
def baseToDec(num_str, base): # suporta até base de valor 36 
  num_str = num_str[::-1]
  num = 0
  for k in range(len(num_str)):
    dig = num_str[k]
    if dig.isdigit():
      dig = int(dig)
    else:  # Assuming its either number or alphabet only
      dig = ord(dig.upper())-ord('A')+10
    num += dig*(base**k)
  return num

In [ ]:
baseToDec("147",8)

103

In [ ]:
tries = 5
time = "0"
names = ""
while time:
  tries -= 1
  name = input('Digite seu nome: ')
  if name == 'name':
    break
  names += name
print(names)

Digite seu nome: iu
Digite seu nome: j
Digite seu nome: name
iuj


**Desafio 038**

Escreva um programa que leia dois números inteiros e compare-os, mostrando na tela uma mensagem:

- O primeiro valor é maior
- O segundo valor é maior 
- Não existe valor maior, os dois são iguais

In [ ]:
n1 = input('Digite o 1º: ')
n2 = input('Digite o 2º: ')
if n1 > n2:
  print('O primeiro valor é maior.')
elif n1 == n2:
  print('\nR: Não existe valor maior, os dois são iguais.')
else:
  print('\nR: O segundo valor é maior')

Digite o 1º: 6
Digite o 2º: 9

R: O segundo valor é maior


**Desafio 039**

Faça um programa que leia o ano de nascimento de um jovem e informe, de acordo com sua idade:

- Se ele ainda vai se alistar ao serviço militar.
- Se é na hora de se alistar.
- Se já passou do tempo de alistamento.

Seu programa também deverá mostrar o tempo que falta ou que passou de prazo

In [ ]:
from datetime import date

year = date.today().year
#print(data_atual)
bornYear = int(input('Say your born year: '))

MILITARYAGE = 18 
age = year-bornYear

militaryTime = age - MILITARYAGE

if militaryTime > 0:
  print('Already passed ',militaryTime, ' year(s) since military register' )
elif militaryTime == 0:
  print('Military register is in this year')
else:
  print(abs(militaryTime), 'year(s) to you have military service register day.')

Say your born year: 2023
19 year(s) to you have military service register day.


In [ ]:
data_atual = date.today()
data_em_texto = '{}/{}/{}'.format(data_atual.day, data_atual.month, data_atual.year)

In [ ]:
data_em_texto

'2/3/2022'

In [ ]:
import datetime

In [ ]:
data_e_hora_atuais = datetime.datetime.now() # truque
data_e_hora_em_texto = data_e_hora_atuais.strftime('%d/%m/%Y')

print(data_e_hora_em_texto)

02/03/2022


In [ ]:
data_e_hora_em_texto = data_e_hora_atuais.strftime('%d/%m/%Y %H:%M')

print(data_e_hora_em_texto)

02/03/2022 16:05


In [ ]:
from datetime import datetime

data_e_hora_em_texto = '17/07/2030 12:55' #14:61'
data_e_hora = datetime.strptime(data_em_texto, '%d/%m/%Y')# %H:%M')
data_e_hora = datetime.strptime(data_e_hora_em_texto, '%d/%m/%Y %H:%M')


In [ ]:
print(data_e_hora)

2030-07-17 12:55:00


**Desafio 40**

Crie um programa que leia duas notas de um aluno e calcule sua média, mostrando uma mensagem no final, de acordo com a média atingida:
- Média abaixo de 5.0:  REPROVADO

- Média entre 5.0 e 6.9: RECUPERAÇÃO

- Média 7.0 ou superior: APROVADO


In [ ]:
grade1 = float(input('Tell your first grade: '))
grade2 = float(input('Tell me your second grade: '))

avg = round((grade1+grade2)/2,1) # se 6.0 e 3.9 dá RECUPERAÇÃO

if avg < 5.0:
  print('REPROVADO')
elif avg < 7.0: 
  print('RECUPERAÇÃO')
else:
  print('APROVADO')
  

Tell your first grade: 7
Tell me your second grade: 9
APROVADO


**Desafio 41**

A Confederação Nacional de Natação precisa de um programa que leia o ano de nascimento de um atleta e mostre sua categoria, de acordo com a idade:

- Até 9 anos: MIRIM
- Até 14 anos: INFANTIL
- Até 19 anos: JUNIOR
- Até 30 anos: SÊNIOR
- Acima: MASTER

In [ ]:
year

2022

In [ ]:
bornYear = int(input('Athlete, say your born year: '))
age = year - bornYear
if age <= 9:
  print('MIRIM')
elif age <= 14:
  print('INFANTIL')
elif age <= 19:
  print('JUNIOR')
elif age <= 30:
  print('SÊNIOR')
else:
  print('MASTER')

Athlete, say your born year: 2018
MIRIM


**DESAFIO 42**

Refaça o Desafio 35 dos triângulos, acrescentando o recurso de mostrar que tipo de triângulo será formado:

* **Equilátero**: todos os lados iguais
* **Isósceles**: dois lados iguais
* **Escaleno**: todos os lados diferentes

In [ ]:
def istriangle(a,b,c):
  # great = a
  # if great<b:
  #   great = b
  # if great < c:
  #   great = c
  # #return great
  # great
  return a < b + c and b < a + c and c < a + b
 

In [ ]:
def triangletype(s1,s2,s3):
  if istriangle(s1,s2,s3):
    if s1 == s2 and s2 == s3:
      print('Equilateral')
    elif s1 == s2 or s1 == s3 or s2 == s3:
      print('Isosceles')
    else:
      print('Scalene')
  else: print('With this measures don\'t is possible build a triangle')

In [ ]:
triangletype(2,3,4)

Scalene


In [ ]:
triangletype(4,3,10)

With this measures don't is possible build a triangle


In [ ]:
def triangletype2(s1,s2,s3):
  if istriangle(s1,s2,s3):
    if s1 == s2:
      if s1 == s3:
        print('Equilateral')
      else: print('Isosceles')
    elif s1 == s3 or s2 == s3:
      print('Isosceles')
    else: print('Scalene') 
  else: print('With this measures don\'t is possible build a triangle')

In [ ]:
triangletype2(1,2,3)

With this measures don't is possible build a triangle


In [ ]:
triangletype2(4,2,5)

Scalene


In [ ]:
triangletype2(3,3,5)

Isosceles


In [ ]:
triangletype(0,0,0)

Equilateral


In [ ]:


#def istriangle(a,b,c): # decomponho para ver qual é o maior e a partir daí vejo se é triangulo
  #if (a>b>c)

In [ ]:
istriangle(2,2,3)

True

In [ ]:
# def istriangle(a,b,c):
#   great(a,b,c)# a > b + c
  

In [ ]:
#greaterSide(3,4,6)

6

**Desafio 043**

Desenvolva uma lógica que leia o peso e a altura de uma pessoa. Calcule seu **IMC** e mostre seu status, de acordo com a tabela abaixo:
- Abaixo de **18.5**: Abaixo do Peso
- Entre **18.5** e **25**: Peso ideal
- **25** até **30**: Sobrepeso
- Acima de **40**:
Obesidade mórbida

In [ ]:
DOWN = 18.5
RIGHT = 25
OVER = 30
OPUP = 40

altura = float(input('Digite a sua altura (em metros): '))
peso = float(input('Digite o seu peso (em kilogramas): '))



IMC = peso/altura**2

if IMC < DOWN:
  print('Below weight\'s')
elif IMC < RIGHT:
  print('Weight optimal ')
elif IMC < OVER:
  print('Above weight\'s')
elif IMC < OPUP:
  print('Morbid obesity')

Digite a sua altura (em metros): 1.89
Digite o seu peso (em kilogramas): 80
Weight optimal 


In [ ]:
IMC

22.395789591556788

**Desafio 044**
Elabore um programa que calcule o valor a ser pago por um produto, considerando o seu **preço normal** e **condição de pagamento**:


*   à vista **dinheiro/cheque: 10%** de desconto
*   à vista no **cartão: 5%** de desconto
*   em até **2* no cartão**: preço normal
*   **3* ou mais** no cartão: [**20%** de juros



In [ ]:
price = float(input('What is the price of your product: '))
payC = str(input('What is payment condition that you choose: ')) # paymentCOndition

if payC == '1':
  price *= 0.95
elif payC == '2':
  price *= 1
elif payC == '3':
  price *= 1.20
elif payC.lower == 'm' or 'c':
  price *= 0.9


print('O preço a ser pago mediante a condição de pagamento é', price)

What is the price of your product: 12
What is payment condition that you choose: c
O preço a ser pago mediante a condição de pagamento é 10.8


**Desafio 045**
Crie um programa que faça o computador **Jokenpô** com voçê.

In [ ]:
import random


In [ ]:
presentation = print('Rock, Paper and Scissors Game')

Rock, Paper and Scissors Game


In [ ]:
gameV =  {1:'rock', 2:'paper', 3:'scissor'}

while True: 
  num = random.randint(0,2) + 1
  human = int(input('Write a number between 1 to 3 which:\n 1 is Rock \n 2 is Paper \n 3 is Scissor\n'))
  if (0 < human < 4):
    break

print(num)
pc = gameV[num]
print(pc)
print(' Jokenpô GAME :\n\n Human move: ', gameV[human],' vs PC move:', pc)#, '\n \n RESULT: ')
dif = human-num #abs(num-human)

print(dif)
if dif == 0:
  print('Draw', pc)
elif dif == 1 or dif == -2:
  print('Vitory')
else:
  print('Defeat') 

#human = gameV[human]

#if 


Write a number between 1 to 3 which:
 1 is Rock 
 2 is Paper 
 3 is Scissor
4
Write a number between 1 to 3 which:
 1 is Rock 
 2 is Paper 
 3 is Scissor
8
Write a number between 1 to 3 which:
 1 is Rock 
 2 is Paper 
 3 is Scissor
-1


In [ ]:
gameV[2]

'paper'

In [ ]:
type(gameV[num])

str

In [ ]:
type(gameV)

dict

In [ ]:
#type(gameV[])

SyntaxError: ignored

In [ ]:
gameV[num]

'scissor'

In [ ]:
# Solução do Gustavo Guanabará
from random import randint
from time import sleep
itens =  ('Pedra', 'Papel', 'Tesoura')
computador = randint(0, 2)
print('''Suas opções:
[ 0 ] PEDRA
[ 1 ] PAPEL
[ 2 ] TESOURA''')
jogador = int(input('Qual é a sua jogada? '))
print('JO')
sleep(1)
print('KEN')
sleep(1)
print('PÔ!!!')
print('-=' * 11)
print('O computador escolheu {}'.format(itens[computador]))
print('Jogador jogou {}'.format(itens[jogador]))
print('-=' * 11)
if computador == 0: # computador jogou P  edra
  if jogador == 0:
    print('EMPATE')
  elif jogador == 1:
    print('JOGADOR VENCE')
  elif jogador == 2:
    print('COMPUTADOR VENCE')
  else:
    print('JOGADA INVÁLIDA!')
elif computador == 1: # computador jogou PAPEL
  if jogador == 0:
    print('COMPUTADOR VENCE')
  elif jogador == 1:
    print('EMPATE')
  elif jogador == 2:
    print('JOGADOR VENCE')
  else:
    print('JOGADA INVÁLIDA!')
elif computador == 1: # computador jogou TESOURA
  if jogador == 0:
    print('JOGADOR VENCE')
  elif jogador == 1:
    print('COMPUTADOR VENCE')
  elif jogador == 2:
    print('EMPATE')
  else:
    print('JOGADA INVÁLIDA!')

Suas opções:
[ 0 ] PEDRA
[ 1 ] PAPEL
[ 2 ] TESOURA
Qual é a sua jogada? 2
-=-=-=-=-=-=-=-=-=-=-=
O computador escolheu Papel
Jogador jogou Tesoura
-=-=-=-=-=-=-=-=-=-=-=
JOGADOR VENCE


In [ ]:
# Estrutura de repetição

# Aula 13

# print('Oi')
# print('Oi')
# print('Oi')
# print('Oi')
# print('Oi')
# print('Oi')


for c in range(0, 6):
  print('Oi')
print('FIM')

# aninhar, colocar uma estrutura dentro da outra

# sintaxe do portinhol e do ...

Oi
Oi
Oi
Oi
Oi
Oi
FIM


In [ ]:
for c in range(6, 0, -1):
  print(c)
print('FIM')

6
5
4
3
2
1
FIM


In [ ]:
n = int(input('Digite um número: '))
#f = int(input)

Digite um número: 56


In [ ]:
s = 0
for c in range(0,4):
  n = int(input('Digite um valor: '))
  s = s + n
print('O somatório de todos os valores foi {}'.format(s))

Digite um valor: 43
Digite um valor: 3
Digite um valor: 7
Digite um valor: 2
O somatório de todos os valores foi 55


# Desafio 046

Faça um programa que mostre na tela **contagem regressiva** para o estouro de fogos de artifício, indo de **10 até 0**, com uma pausa de **1 segundo** entre eles.


In [ ]:
import time

for c in range(10,-1, -1):
  print(c)
  time.sleep(1)

10
9
8
7
6
5
4
3
2
1
0


In [1]:
# Answer by Gustavo Guanabará
from time import sleep
for cont in range(10, -1, -1): # (0,11,1)
  print(cont)
  sleep(0.5)
print('BUM! BUM! POOOW!')

10
9
8
7
6
5
4
3
2
1
0
BUM! BUM! POOOW!


# Desafio 047

Crie um programa que mostre na tela **todos os números pares** que estão no intervalo de **1** e **50**

In [ ]:
for c in range(0,50):
  if c%2==0:
    print(c)

0
2
4
6
8
10
12
14
16
18
20
22
24
26
28
30
32
34
36
38
40
42
44
46
48


In [5]:
#Solved by Gustavo Guanabará
for n in range(1, 51, 2):
  # print('.', end='')
  # if n % 2 == 0: #print(n, end=' ')
    #print(n, end=' ')
  print(n, end=' ') # tempo do processor métode do tempo! 
print('Acabou')

1 3 5 7 9 11 13 15 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 Acabou


# Desafio 048

Faça um programa que calcule a **soma** entre todos os **número ímpares** que são **múltiplos de três** e que se encontram no intervalo de **1** até **500**


In [ ]:
sum = 0
for c in range(1, 500):
  if c %3 == 0 and c %2 ==1:
    sum += c
    print(c)
print(sum)
# print('termo médio vezes 2, ou os dois do meio se guardar num vetor, -1 ')



3
9
15
21
27
33
39
45
51
57
63
69
75
81
87
93
99
105
111
117
123
129
135
141
147
153
159
165
171
177
183
189
195
201
207
213
219
225
231
237
243
249
255
261
267
273
279
285
291
297
303
309
315
321
327
333
339
345
351
357
363
369
375
381
387
393
399
405
411
417
423
429
435
441
447
453
459
465
471
477
483
489
495
20667


In [12]:
# Solution by Gustavo Guanabará

soma = cont = 0
for c in range(1, 501, 2): # números ímpares!
  if c % 3 == 0:
    cont = cont + 1
    soma = soma + c
   # print(c, end=' ')
print('A soma de todos os {} valores solicitados é {}'.format(cont, soma))

A soma de todos os 83 valores solicitados é 20667


# Desafio 049

Refaça o **DESAFIO 009**, mostrando a **tabuada** de um número que o usuário escolher, só que agora utilizando um **laço for**.


In [ ]:
tabuada = int(input('Digite o número da tabuada que deseja: '))
print('\n Tabuada do {}'.format(tabuada))

print('\n')
print('-' * 39)

for c in range(0, 11):
  print('{} * {} = {}'.format(tabuada, c, tabuada*c))


Digite o número da tabuada que deseja: 5

 Tabuada do 5


---------------------------------------
5 * 0 = 0
5 * 1 = 5
5 * 2 = 10
5 * 3 = 15
5 * 4 = 20
5 * 5 = 25
5 * 6 = 30
5 * 7 = 35
5 * 8 = 40
5 * 9 = 45
5 * 10 = 50


In [14]:
# Answer by Gustavo Guanabará

num = int(input('Digite um número para ver sua tabuada: '))
for c in range(1,11):
  print('{} * {:2} = {}'.format(num, c, num*c))

Digite um número para ver sua tabuada: 7
7 *  1 = 7
7 *  2 = 14
7 *  3 = 21
7 *  4 = 28
7 *  5 = 35
7 *  6 = 42
7 *  7 = 49
7 *  8 = 56
7 *  9 = 63
7 * 10 = 70


# Desafio 050

Desenvolva um programa que leia **seis números inteiros** e mostre a soma apenas daqueles que forem **pares**. Se o valor digitado for **ímpar**, desconsidere-o.

In [ ]:
sum = 0
num={}
for c in range(1, 7):
  num[c] = int(input('Digit the number that you pretend: '))
  if num[c]%2 == 0:
    sum += num[c]
print(sum)


Digit the number that you pretend: 22
Digit the number that you pretend: 1
Digit the number that you pretend: 3
Digit the number that you pretend: 6
Digit the number that you pretend: 8
Digit the number that you pretend: 10
46


In [19]:
# Answered by Gustavo Guanabará

soma = cont = 0

for c in range(1,7):
  num = int(input('Digite o {} valor: '.format(c)))
  if num % 2 == 0:
    soma = soma + num
    cont += 1
print('Você informou {} números e a soma foi {}'.format(cont, soma))

Digite o 1 valor: 6
Digite o 2 valor: 9
Digite o 3 valor: 11
Digite o 4 valor: 12
Digite o 5 valor: 13
Digite o 6 valor: 8
Você informou 3 números e a soma foi 26


# Desafio 051

Desenvolve um programa que leia o **primeiro termo** e a **razão** de uma **PA**. No final, mostre os 10 primeiros termos dessa progressão.

In [38]:
# PA é progressão aritmética, então vem da soma. razão de 4 e primeiro termo seja 2, P1 será 6, P2 -> 10 etc

# base cases
prog = {}
prog[0] = int(input("Digit firST term: "))
ratio = int(input("Please, digit the ratio of formula: "))

print('0º -> {}'.format(prog[0]))
for c in range(1,11):
  prog[c] = prog[c-1] + ratio
  print('{}º -> {}'.format(c, prog[c]))

Digit firST term: 7
Please, digit the ratio of formula: 8
0º -> 7
1º -> 15
2º -> 23
3º -> 31
4º -> 39
5º -> 47
6º -> 55
7º -> 63
8º -> 71
9º -> 79
10º -> 87


In [39]:
# Solução Híbrida, feita à posteriori

primeiro = int(input('Primeiro termo: '))
razao = int(input('Razão: '))
#décimo = primeiro + (10-1)* razao
enésimo = int(input('How until term you want?: '))
for c in range(primeiro, primeiro + enésimo*razao, razao): # se o step for a razão fica muito acessível ...
  print('{} '.format(c), end=' ➙ ')
print('ACABOU')

Primeiro termo: 5
Razão: 6
How until term you want?: 3
5  ➙ 11  ➙ 17  ➙ ACABOU


In [40]:
# Solution by Gustavo Guanabara

# tiramos o elefante branco da sala

primeiro = int(input('Primeiro termo: '))
razao = int(input('Razão: '))
décimo = primeiro + (10-1)* razao
for c in range(primeiro, décimo + razao, razao): # se o step for a razão fica muito acessível ...
  print('{} '.format(c), end=' ➙ ')
print('ACABOU')

Primeiro termo: 25
Razão: 5
25  ➙ 30  ➙ 35  ➙ 40  ➙ 45  ➙ 50  ➙ 55  ➙ 60  ➙ 65  ➙ 70  ➙ ACABOU


# Desafio 052

 Faça um programa que leia um **número inteiro** e diga se ele é ou não um **número primo**.
 


In [26]:
import math
from math import sqrt
num = int(input('Digit a number: '))
primo = True 
if num == 0 or num == 1:
  primo = False
for c in range(2, int(math.sqrt(num))):
  primo = False
  break
print('É primo: {}'.format(primo))


Digit a number: 1
É primo: False


In [48]:
# Answered by Gustavo Guanabará

num = int(input('Digite um número: '))
tot = 0
for c in range(1, num +1):
  if num % c == 0:
    print('\033[33m', end='')
    tot += 1
  else:
    print('\033[31m', end='')
  print('{} '.format(c), end='')
print('\n\033[mO número {} foi divisível {} vezes'.format(num, tot))
if tot == 2:
  print('E por isso ele É PRIMO!')
else:
  print('É por isso ele NÃO É PRIMO!')
 # print('{} '.format(c), end='')

Digite um número: 67
1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 
O número 67 foi divisível 2 vezes
E por isso ele É PRIMO!


# Desafio 053

Crie um programa que leia uma **frase** qualquer e diga se ela é um **palídromo**,
desconsiderando os espaços.

Ex:

APOS A SOPA

A SACADA DA CASA

A TORRE DA DERROTA

 O LOBO AMA O BOLO

ANOTARAM A DATA DA MARATONA

In [9]:
import sys

phrase = str(input('Digit the phrase: '))
print(len(phrase))
phrase = phrase.strip()
phrase = phrase.replace(" ", "")
mid = int(len(phrase)/2)
print(phrase)

for c in range(0, mid):
  if phrase[c] != phrase[len(phrase)-c-1]:
    #print('Not palindrome')
    sys.exit('Not palindrome')
print('It\'s PALINDROME')

Digit the phrase: O LOBO AMA O BOLO
17
OLOBOAMAOBOLO
It's PALINDROME


In [2]:
# Answered by Gustavo Guanabara

# Tradicional version

frase = str(input('Digite uma frase: ')).strip().upper()
palavras = frase.split()
junto = ''.join(palavras) #'*'.join(palavras)
#print(junto)#palavras)
inverso = ''
inverso = junto[::-1] # fatiamento ou macete do fatiamento do python
''' for letra in range(len(junto)-1, -1, -1): # vai para o último index como começa em zero, temos de tirar 1 ..., depois vai até ao início que é 0 mas como é exclusivo .. temos de ir 1a além dessa: daí o -1,  como vai no sentido inverso do habitual then this!
   #print(junto[letra])
   inverso += junto[letra] '''
print('O inverso de {} é {}'.format(junto, inverso))
if inverso == junto:
  print('Temos um palíndromo!')
else:
  print('A frase digitada não é um palíndromo!')
# print('Você digitou a frase {}'.format(#junto))

Digite uma frase: Hello Mocho!
O inverso de HELLOMOCHO! é !OHCOMOLLEH
A frase digitada não é um palíndromo!


In [25]:
type(int(len(phrase)/2))

int

# Desafio 054

Crie um programa que leia o **ano de nascimento** de **sete pessoas**. No final, mostre quantas pessoas ainda não atingiram a maioridade e quantas já são maiores.

In [25]:
def clean_date_of_birth(self):
    date = self.cleaned_data['date_of_birth']
    if date > datetime.date.today():  # 🖘 raise error if greater than
        raise forms.ValidationError("The date cannot be in the future!")
    return date

In [3]:
from datetime import date
adults = 0
tries = childs = 0
ano = {}
actualYear = date.today().year
print(actualYear)
for c in range(0,7):
  ano[c] = int(input('Digit the bornYear: '))
  while ano[c] > actualYear:
    ano[c] = int(input('Digit the bornYear: '))
    tries += 1
  # while (date.today() < ano[c]):
  #   ano[C] = int(input('Digit the bornYear: '))
  print(ano[c])
  if actualYear - ano[c] >= 21:
    adults += 1
  else:#if actualYear - ano[c]< 0:  # garantir que não tem datas futuras... 
    childs += 1
print('Number of Babys: {} \n Number of Adults: {}'.format(childs, adults))


2022
Digit the bornYear: 2024
Digit the bornYear: 2022
2022
Digit the bornYear: 2021
2021
Digit the bornYear: 2023
Digit the bornYear: 2021
2021
Digit the bornYear: 2000
2000
Digit the bornYear: 2222
Digit the bornYear: 221
221
Digit the bornYear: 2012
2012
Digit the bornYear: 2013
2013
Number of Babys: 5 
 Number of Adults: 2


In [22]:
#install python 3.9
!sudo apt-get update -y
!sudo apt-get install python3.9

#changes alternatives
#!python --version


Get:1 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease [1,581 B]
Get:2 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Get:3 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]
Get:4 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Hit:7 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [817 kB]
Get:9 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:10 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ Packages [85.6 kB]
Hit:11 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Hit:13 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bioni

In [25]:
# changes alternatives
!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.7 1
!sudo update-alternatives --install /usr/bin/python3 python3 /usr/bin/python3.9 2

update-alternatives: using /usr/bin/python3.9 to provide /usr/bin/python3 (python3) in auto mode


In [27]:
from datetime import date
adults = 0
childs = 0
ano = {}
actualYear = date.today().year
print(actualYear)
for c in range(0,7):
  while actualYear < (bornYear := int(input('Digit the bornYear: '))): 
  # while (date.today() < ano[c]):
  #   ano[C] = int(input('Digit the bornYear: '))
  print(ano[c])
  if actualYear - ano[c] >= 21:
    adults += 1
  else:#if actualYear - ano[c]< 0:  # garantir que não tem datas futuras... 
    childs += 1
print('Number of Babys: {} \n Number of Adults: {}'.format(childs, adults))

SyntaxError: ignored

In [26]:
!python --version

Python 3.9.13


In [9]:
# Answered by Gustavo Guanabará

from datetime import date
atual = date.today().year
totmaior = 0
totmenor = 0
for pess in range(1,8):
  nasc = int(input('Em que ano a {}ª pessoa nasceu? '.format(pess)))
  idade = atual - nasc
  if idade >= 21:
    #print('Essa pessoa é maior')
    totmaior += 1
  else:
    #print('Essa pessoa é menor')
    totmenor += 1
print('Ao todo tivemos {} pessoas maiores de idade'.format(totmaior))
print('Ao todo tivemos {} pessoas menores de idade'.format(totmenor))

Em que ano a 1ª pessoa nasceu? 3
Em que ano a 2ª pessoa nasceu? 5
Em que ano a 3ª pessoa nasceu? 6
Em que ano a 4ª pessoa nasceu? 7
Em que ano a 5ª pessoa nasceu? 6
Em que ano a 6ª pessoa nasceu? 5674
Em que ano a 7ª pessoa nasceu? 4365
Ao todo tivemos 5 pessoas maiores de idade
Ao todo tivemos 2 pessoas menores de idade


# Desafio 055

Faça um programa que leia o **peso** de **cinco pessoas**. No final, mostre qual foi o **maior** e o **menor** peso lidos.

In [29]:
weight = {}
min = 1000
max = 0 
for c in range(0, 5): 
  weight[c] = int(input('Digit the person weight: '))
  if weight[c] < min:
    min = weight[c]
  if max < weight[c]: # nada a ver , mas se todos iguais não atualiza ...
    max = weight[c]
print('Os pesos máx. = {} e min. = {}'.format(max, min))

Digit the person weight: 120
Digit the person weight: 56
Digit the person weight: 78
Digit the person weight: 75
Digit the person weight: 90
Os pesos máx. = 56 e min. = 120


In [35]:
import numpy as np

In [43]:
list1 = [67, 53, 88, 39, 102]
# weight = {67, 53, 88, 39, 102}
weight = np.array(list1)
min = max = weight[0]
for c in range(0, len(weight)):
  if min > weight[c]:
    min = weight[c]
  elif weight[c] > max:
    max = weight[c]
print('The weights maximum is {} and the minimum is {}'.format(max, min)) 


The weights maximum is 102 and minimum is 39


In [44]:
weight[0]

67

In [12]:
# Solution by Gustavo Guanabara

# Análise de dados, área de guarde importância no momentoi (totaliza, outra count .. ou diz o maior e menor [extremos] de uma lista)
maior = 0
menor = 0
for p in range(1,6):
  peso = float(input('Peso da {}ª pessoa: '.format(p)))
  if p == 1:
    maior = peso
    menor = peso
  else:
    if peso > maior:
      maior = peso
    if peso < menor:
      menor = peso
print('O maior peso lido foi de {}Kg'.format(maior))
print('O menor peso lido foi de {}Kg'.format(menor))

Peso da 1ª pessoa: 45
Peso da 2ª pessoa: 75
Peso da 3ª pessoa: 98
Peso da 4ª pessoa: 34
Peso da 5ª pessoa: 86
O maior peso lido foi de 98.0Kg
O menor peso lido foi de 34.0Kg


In [11]:
p

5

# Desafio 056

Desenvolva um programa que leia o **nome**, **idade** e **sexo** de **4 pessoas**. No final do programa, mostre:



*   A **média de idade** do grupo.
*   Qual é o nome do homem **mais velho**.
*   Quantas mulheres têm **menos de 20 anos**.



In [17]:
names = {}
ages = {}
#sex = []
sex = {}
maXman = sum = count = max = 0
maXmanI = -1 # saber o nome do homem mais velho..., implica saber o index do man
S = 4
for c in range(0, S):
  names[c] = str(input(' Digit person name: '))
  ages[c] = int(input('Digit person age: '))
  sex[c] = str(input('Please, digit your sex: F or M?: '))
  sum += ages[c]
  if sex.upper().strip() == 'F' and ages[c]<20:
  #if sex.strip() == 'F' and ages[c]<20:
    count += 1
  if sex.upper() == 'M':
    if (max<ages[c]):
      max = ages[c]
      maXmanI = c
mean = sum / S
print('A média de idades é: {}, o homem mais velho é: {} e\n existem {} mulheres com menos de 20 anos!'.format(mean,names[maXmanI], count))     



 Digit person name: Gago
Digit person age: 45
Please, digit your sex: F or M?: M


AttributeError: ignored

In [47]:
type(sex)

dict

In [30]:
# Answer by Gustavo Guanabara
somaidade = 0
meidaidade = 0
maioridadehomem = totmulher20 = 0
nomevelho = ''
for p in range(1, 5):
  print('------ {}ª PESSOA ------'.format(p))
  nome = str(input('Nome: ')).strip()
  idade = int(input('Idade: '))
  sexo = str(input('Sexo [M/F]: ')).strip()
  somaidade += idade
  if p == 1 and sexo in 'Mm':
    maioridadehomem = idade
    nomevelho = nome 
  if sexo in 'Mm' and idade > maioridadehomem:
    maioridadehomem = idade
    nomevelho = nome
  if sexo in 'Ff' and idade < 20: #interessante, cadeia de string para definir.., evita a utilização de upper etc
    totmulher20 += 1
mediaidade = somaidade / 4
print('A média de idade do grupo é de {}.'.format(mediaidade))
print('O homem mais velho tem {} anos e se chama {}.'.format(maioridadehomem, nomevelho))
print('Ao todo são {} mulheres com menos de 20 anos'.format(totmulher20))

------ 1ª PESSOA ------
Nome: Ju
Idade: 12
Sexo [M/F]: F
------ 2ª PESSOA ------
Nome: Gasd
Idade: 30
Sexo [M/F]: M
------ 3ª PESSOA ------
Nome: ffg
Idade: 19
Sexo [M/F]: F
------ 4ª PESSOA ------
Nome: hugo
Idade: 21
Sexo [M/F]: F
A média de idade do grupo é de 20.5.
O homem mais velho tem 30 abnos e se chama Gasd.
Ao todo são 2 mulheres com menos de 20 anos


In [ ]:
# For : estrutura de repetição com variável de controlo <- qdo tem limite
# While : estrutura de repetição (não existe um padrão, está muito irregular!)
# não existem mais rápido entre os dois

In [1]:
# Aula 14

for c in range(1,10): # Quando não sei o limite não dá para usar o FOR
  print(c)
print('Fim')



1
2
3
4
5
6
7
8
9
Fim


In [2]:
c = 1
while c < 10: 
  print(c)
  c = c + 1
print('Fim')

1
2
3
4
5
6
7
8
9
Fim


In [3]:
n = 1
while n != 0:  # FLAG ou condição de parada
  n = int(input('Digite um valor: '))
print('Fim')

Digite um valor: 5
Digite um valor: 8
Digite um valor: 0
Fim


In [4]:
r = 'S'
while r == 'S': # faz laços indetermininadamente
  n = int(input('Digite um valor: '))
  r = str(input('Quer continuar? [S/N] ')).upper()  
print('Fim')

Digite um valor: 45
Quer continuar? [S/N] S
Digite um valor: 323
Quer continuar? [S/N] 7
Fim


In [ ]:
n = 1
par = impar = 0
while n != 0:
  n = int(input('Digite um valor: '))
  if n != 0: # gambiarra
    if n % 2 == 0:
      par += 1 
    else:
      impar += 1
print('Voçê digitou {} números pares e {} números ímpares!'.format(par, impar))

# Desafio 57

Faça um programa que leia o **sexo** de uma pessoa, mas só aceite os valores 'M' ou 'F'. 
Caso esteja errado, peça a digitação novamente até ter um valor correto.

In [6]:
gender = str(input('Say your sex: '))  
while gender not in 'Ff' and gender not in 'mM':
  gender = str(input('Say your sex: '))  

Say your sex: g
Say your sex: M


**Desafio 58**

Melhore o jogo do **Desafio 028** onde o computador vai "pensar" em um **número entre 0 e 10**. Só que agora o jogador vai tentar adivinhar até acertar, mostrando no final quantos palpites foram necessários para vencer.

In [7]:
import random
count = 1
num = random.randint(0, 11)
unum = int(input('Digite um número entre 0 e 10: '))
while num != unum :
  unum = int(input('Digite um número entre 0 e 10: '))
  count += 1
print('The player guess the num in {} tries'.format(count))

Digite um número entre 0 e 10: 6
Digite um número entre 0 e 10: 9
Digite um número entre 0 e 10: 1
Digite um número entre 0 e 10: 0
Digite um número entre 0 e 10: 3
Digite um número entre 0 e 10: 2
The player guess the num in 6 tries


# Desafio 059

Crie um programa que leia **dois valores** e mostre um **menu** na tela:



[1]    Somar

[2]    Multiplicar

[3]    Maior

[4]    Novos números

[5]    Sair do programa


Seu programa deverá realizar a operação solicitada em cada caso.


In [ ]:
def readNums():
  a = int(input('Digit a number: '))
  b = int(input('Digit other number: '))
  return a, b

In [18]:
# def ciclo(option):
#     if option == 1:
#       print('Sum: {} + {} = {}'.format(num1, num2, num1+num2))
#     elif option == 2:
#       print('Multiplication: {} * {} = {}'.format(num1, num2, num1*num2))
#     elif option == 3:
#       print('Greater is {}'.format(max(num1, num2)))
#     elif option == 4:
#       num1, num2 = readNums()
#     elif option == 5:
#       print('The End!')
#       break

SyntaxError: ignored

In [26]:
option = 3.5
while True:
  if option == 3.5:
    num1, num2 = readNums()
  else:
    print('''\nBem Vindo ao Menu! 
          \nDigite uma das opções:\n
            [1] Somar
            [2] Multiplicar
            [3] Maior
            [4] Novos números
            [5] Sair do programa\n\n''')


    
            
  while option > 6 and option < 0:
    option = int(input('Say a option: '))
  if option == 1:
    print('Sum: {} + {} = {}'.format(num1, num2, num1+num2)) 
  elif option == 2:
    print('Multiplication: {} * {} = {}'.format(num1, num2, num1*num2)) 
  elif option == 3:
    print('Greater is {}'.format(max(num1, num2))) 
  elif option == 4: 
    num1, num2 = readNums() 
  elif option == 5: 
    print('The End!') 
    break
  option = int(input('Say a option: '))


Digit a number: 4
Digit other number: 3
Say a option: 1

Bem Vindo ao Menu! 
          
Digite uma das opções:

            [1] Somar
            [2] Multiplicar
            [3] Maior
            [4] Novos números
            [5] Sair do programa


Sum: 4 + 3 = 7
Say a option: 5

Bem Vindo ao Menu! 
          
Digite uma das opções:

            [1] Somar
            [2] Multiplicar
            [3] Maior
            [4] Novos números
            [5] Sair do programa


The End!


In [17]:
# num1, num2 = readNums()
# # option = 1
# # while (option > 0):
# #   option = int(input('Say a option: '))
# option = 4
# while (option > 5): # necesaário se quiser garantir que quando vai fora do scope de 1 a cinco.., faz repetir a opção
#   option = int(input('Say a option: '))
#   if option > 0 and option < 6: # ao invés do while
#     option = int(input('Say a option: '))
#     if option == 1:
#       print('Sum: {} + {} = {}'.format(num1, num2, num1+num2))
#     elif option == 2:
#       print('Multiplication: {} * {} = {}'.format(num1, num2, num1*num2))
#     elif option == 3:
#       print('Greater is {}'.format(max(num1, num2)))
#     elif option == 4:
#       num1, num2 = readNums()
#     elif option == 5:
#       print('The End!')
#       break
#     # else:
#     #   print('The End!')
#     #   break

Digit a number: 56
Digit other number: 12
Say a option: 2
Multiplication: 56 * 12 = 672
Say a option: 1
Sum: 56 + 12 = 68
Say a option: 3
Greater is 56
Say a option: 4
Digit a number: 43
Digit other number: 78
Say a option: 1
Sum: 43 + 78 = 121
Say a option: 7


# Desafio 60

Faça um programa que leia um **número** qualquer e mostre o seu **fatorial**.

Ex:
5! = 5 * 4 * 3 * 2 * 1 = 120 





In [59]:
def fatorial(n):
  if n == 1:
    print('{} \n='.format(n))
    return 1
  else:
    print('{} \n*'.format(n , end = ' '))
    return (n * fatorial(n-1))

In [77]:
fat = int(input('Say fatorial number that you wish: '))
print('{} = {}!'.format(fatorial(fat), fat))

Say fatorial number that you wish: 10
10 
*
9 
*
8 
*
7 
*
6 
*
5 
*
4 
*
3 
*
2 
*
1 
=
3628800 = 10!


In [78]:
def fatorial2 (fat): 
  result = 1 # In multiplication, identity element is 1, division to
  while fat > 0:
    print('{} \n*'.format(fat , end = ' '))
    result *= fat
    fat -= 1
  return result



In [79]:
print('{} = {}!'.format(fatorial2(fat), fat))

10 
*
9 
*
8 
*
7 
*
6 
*
5 
*
4 
*
3 
*
2 
*
1 
*
3628800 = 10!


In [80]:
def fatorial3 (fat): 
  result = 1 # In multiplication, identity element is 1, division to
  for n in range(1,fat+1):
    print('{} \n*'.format(n , end = ' '))
    result *= n
  return result

In [81]:
def fatorial4 (fat): 
  result = 1 # In multiplication, identity element is 1, division to
  for n in range(fat,0, -1):
    print('{} \n*'.format(n , end = ' '))
    result *= n
  return result

In [82]:
print('{} = {}!'.format(fatorial3(fat), fat))

1 
*
2 
*
3 
*
4 
*
5 
*
6 
*
7 
*
8 
*
9 
*
10 
*
3628800 = 10!


In [83]:
print('{} = {}!'.format(fatorial4(fat), fat))

10 
*
9 
*
8 
*
7 
*
6 
*
5 
*
4 
*
3 
*
2 
*
1 
*
3628800 = 10!


# Desafio 061

Refaça o **Desafio 051**, lendo o **primeiro termo** e a **razão** de uma **PA**, mostrando os **10 primeiros termos** da progressão usando a estrutura **while**

In [92]:
pav = basis = float(input("Digit first term: \n"))

ratio = float(input("Say ration that you want: \n"))

count = 0

print('The first 10 term of PA is: \n')
while count < 10:
  pav += ratio
  count += 1
  print('{:.3f}'.format(pav))

Digit first term: 
1.51
Say ration that you want: 
0.1
The first 10 term of PA is: 

1.610
1.710
1.810
1.910
2.010
2.110
2.210
2.310
2.410
2.510
